### DISCUSSION TITLE
Help needed to debug #1334

### DISCUSSION THREAD

**Discussion by tmssngr** [eclipse-platform/eclipse.platform.swt#1338] [General]
How can I debug SWT in our applications to find out why no SWT.Expand event is sent when pressing the right arrow key on a tree node?
https://github.com/eclipse-platform/eclipse.platform.swt/issues/1334

**Comment by iloveeclipse** [comment]
Is the question or bug about "in **our** application we don't see the event, but in others the event is there" ?
Or is this about "how to debug SWT in **our** application"?
Or is this about "how to debug **any** SWT application"?

**Comment by tmssngr** [comment]
In our applications the SWT.Expand event is NOT sent, but in a tiny Tree test snippet it is sent. Where should I set breakpoints that are triggered before the expansion happens after having pressed the cursor-right key? The question is "how to debug *this problem* in our application", so I can find out what triggers it, so I get empowered to create a reproducer test snippet.

**Comment by iloveeclipse** [comment]
I would debug it in a (working) tiny SWT test snippet and put breakpoints in all code backwards from the place where event is received.
Then I would debug my application and see which of the breakpoints isn't being hit, and look around why.

**Comment by tmssngr** [comment]
That's what I've already tried with this snippet:
```java
import org.eclipse.swt.*;
import org.eclipse.swt.layout.*;
import org.eclipse.swt.widgets.*;

public class Snippet202 {

    public static void main(String[] args) {
        Display display = new Display();
        final Shell shell = new Shell(display);
        shell.setText("Snippet 202");
        shell.setLayout (new FillLayout());
        final Tree tree = new Tree(shell, SWT.VIRTUAL | SWT.BORDER);
        tree.addListener(SWT.SetData, event -> {
            TreeItem item = (TreeItem)event.item;
            TreeItem parentItem = item.getParentItem();
            String text = null;
            if (parentItem == null) {
                text = "node "+tree.indexOf(item);
            } else {
                text = parentItem.getText()+" - "+parentItem.indexOf(item);
            }
            item.setText(text);
            item.setItemCount(10);
        });
        tree.setItemCount(20);
        tree.addListener(SWT.Expand, event -> System.out.println("expanded"));
        shell.setSize(400, 300);
        shell.open();
        while (!shell.isDisposed ()) {
            if (!display.readAndDispatch ()) display.sleep ();
        }
        display.dispose ();
    }
}
```
but it does not help. Having set a break-point in the listener gives following call stack:
```
	at Snippet202.lambda$main$1(Snippet202.java:26)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:91)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5855)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1617)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1643)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1626)
	at org.eclipse.swt.widgets.Tree.gtk_test_expand_row(Tree.java:2620)
	at org.eclipse.swt.widgets.Widget.windowProc(Widget.java:2624)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:6168)
	at org.eclipse.swt.internal.gtk3.GTK3.gtk_main_do_event(Native Method)
	at org.eclipse.swt.widgets.Display.eventProc(Display.java:1598)
	at org.eclipse.swt.internal.gtk3.GTK3.gtk_main_iteration_do(Native Method)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4514)
	at Snippet202.main(Snippet202.java:30)
```
and I already wrote in #1334, that `Tree.gtk_test_expand_row` is not invoked in our applications.

**Comment by iloveeclipse** [comment]
Just a wild guess: I assume in your application the tree is nested in many other components/scrollbars/panes etc. Try to create similar widget hierarchy in the simple snippet, may be some of the parent SWT widgets "eats" the event before it is arrived at the tree.

**Comment by tmssngr** [comment]
Shame on me - it was an interfering key listener. I've commented all non-essential code in our Tree-wrapping class; then it worked.

Thanks for the tips.

**Comment by iloveeclipse** [comment]
Sometimes one just need to discuss the problem with someone else, even if the someone else has no clue :-)
